### PR TITLE
Go 1.6

### LINKED ISSUE #1730: Update to Go 1.6

### PR DESCRIPTION
**Pull request by titanous**
Closes #1730

@lmars There was a data race that I fixed in 690034d1ee527fe63fe6a89421112fb0939104ae but it revealed a logical race after fixing, please take a look when you have a minute and push a patch.

### CODE REVIEW

**Comment by titanous** [comment]
@josephglanville This appeared in a recent run, not sure if it's a bug on this branch or in master:

```text
t=2016-03-11T21:11:14+0000 lvl=info msg="request started" component=mariadb-api req_id=f7830fa0-bb1c-4ae0-94c0-f5acb6a7e4d9 method=POST path=/databases client_ip=100.100.91.5
t=2016-03-11T21:11:14+0000 lvl=eror msg="Error 1064: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '2e1ebd37477f5468b36a67e1cd881269' at line 1" component=mariadb-api req_id=f7830fa0-bb1c-4ae0-94c0-f5acb6a7e4d9
```

https://ci.flynn.io/builds/20160311210045-460b9341

**Comment by lmars** [comment]
@titanous scheduler test fixed: 194a4f8

File: util/packer/ubuntu-14.04/scripts/install.sh [hunk 1]
```diff
@@ -220,7 +220,7 @@ bump_libvirt_start_timeout() {
 install_go() {
   cd /tmp
   curl https://godeb.s3.amazonaws.com/godeb-amd64.tar.gz | tar xz
-  ./godeb install 1.4.3
+  ./godeb install 1.6.0
```
Comment (lmars) [inline util/packer/ubuntu-14.04/scripts/install.sh]: This should be `./godeb install 1.6`:

```
$ ./godeb install 1.6.0
error: version 1.6.0 not available at http://golang.org/dl/

$ ./godeb install 1.6  
processing https://storage.googleapis.com/golang/go1.6.linux-amd64.tar.gz
...
```

**Comment by josephglanville** [comment]
LGTM

**Comment by titanous** [comment]
There is one known outstanding race: Jobs in the `flynn-host` state are currently modified concurrently with reads. While starting a job, its environment variables are modified by the backend, this map may be read concurrently by the JSON serializer.

**Comment by titanous** [comment]
Another race panic:

```text
t=2016-04-09T02:03:19+0000 lvl=info msg="request started" component=host req_id=a425d3bc-c0f9-409f-b61c-2ecc32b85db1 method=PUT path=/host/jobs/50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084 client_ip=10.69.14.4
t=2016-04-09T02:03:19+0000 lvl=info msg="decoding job" app=host pid=1130 host.id=50cb2da0 fn=AddJob job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084
t=2016-04-09T02:03:19+0000 lvl=info msg="acquiring state database" app=host pid=1130 host.id=50cb2da0 fn=AddJob job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084
t=2016-04-09T02:03:19+0000 lvl=info msg="request completed" component=host req_id=a425d3bc-c0f9-409f-b61c-2ecc32b85db1 status=200 duration=2.158856ms
t=2016-04-09T02:03:19+0000 lvl=info msg="running job" app=host pid=1130 host.id=50cb2da0 fn=AddJob job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084
now=2016-04-09T02:03:19+0000 at=start job.artifact.uri="https://example.com?name=flynn/test-apps&id=3064d189d78399cb821514717f33a054e2437735c62ff06bf990426e562f8d41" fn=run job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084 backend=libvirt-lxc job.cmd="[sh -c while true; do echo I like to print; sleep 1; done]"
now=2016-04-09T02:03:19+0000 backend=libvirt-lxc fn=run job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084 at=pull_image
now=2016-04-09T02:03:19+0000 backend=libvirt-lxc fn=run job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084 at=read_config
now=2016-04-09T02:03:19+0000 backend=libvirt-lxc fn=run job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084 at=checkout
now=2016-04-09T02:03:19+0000 fn=run job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084 at=mount backend=libvirt-lxc
fatal error: concurrent map read and map write
now=2016-04-09T02:03:19+0000 backend=libvirt-lxc fn=run job.id=50cb2da0-696d863d-6f00-4a18-bd03-4c72d9e77084 at=write_config

goroutine 2375 [running]:
runtime.throw(0xd2bf60, 0x21)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/runtime/panic.go:530 +0x90 fp=0xc820799200 sp=0xc8207991e8
runtime.mapaccess2(0xa50be0, 0xc820772ff0, 0xc8207bba80, 0xc, 0x5fdf1e)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/runtime/hashmap.go:343 +0x5a fp=0xc820799248 sp=0xc820799200
reflect.mapaccess(0xa50be0, 0xc820772ff0, 0xc8207bba80, 0xc820772ff0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/runtime/hashmap.go:993 +0x35 fp=0xc820799278 sp=0xc820799248
reflect.Value.MapIndex(0xa50be0, 0xc82069e888, 0x195, 0xa52060, 0xc8207bba80, 0x98, 0x0, 0x0, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/reflect/value.go:1041 +0x14a fp=0xc820799300 sp=0xc820799278
encoding/json.(*mapEncoder).encode(0xc820022d60, 0xc820ace6e0, 0xa50be0, 0xc82069e888, 0x195, 0x300)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:622 +0x2ec fp=0xc8207993f0 sp=0xc820799300
encoding/json.(*mapEncoder).(encoding/json.encode)-fm(0xc820ace6e0, 0xa50be0, 0xc82069e888, 0x195, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:632 +0x51 fp=0xc820799428 sp=0xc8207993f0
encoding/json.(*structEncoder).encode(0xc8200db8c0, 0xc820ace6e0, 0xc3e4a0, 0xc82069e850, 0x199, 0x200)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:587 +0x2c4 fp=0xc8207995d0 sp=0xc820799428
encoding/json.(*structEncoder).(encoding/json.encode)-fm(0xc820ace6e0, 0xc3e4a0, 0xc82069e850, 0x199, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:601 +0x51 fp=0xc820799608 sp=0xc8207995d0
encoding/json.(*structEncoder).encode(0xc8200db830, 0xc820ace6e0, 0xc00a20, 0xc82069e800, 0x199, 0x5fdf00)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:587 +0x2c4 fp=0xc8207997b0 sp=0xc820799608
encoding/json.(*structEncoder).(encoding/json.encode)-fm(0xc820ace6e0, 0xc00a20, 0xc82069e800, 0x199, 0xc82069e800)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:601 +0x51 fp=0xc8207997e8 sp=0xc8207997b0
encoding/json.(*ptrEncoder).encode(0xc8203b4310, 0xc820ace6e0, 0xb4fc00, 0xc820a0ef00, 0x196, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:709 +0xea fp=0xc820799838 sp=0xc8207997e8
encoding/json.(*ptrEncoder).(encoding/json.encode)-fm(0xc820ace6e0, 0xb4fc00, 0xc820a0ef00, 0x196, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:714 +0x51 fp=0xc820799870 sp=0xc820799838
encoding/json.(*structEncoder).encode(0xc8200db800, 0xc820ace6e0, 0xc248c0, 0xc820a0ef00, 0x199, 0x5fdf00)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:587 +0x2c4 fp=0xc820799a18 sp=0xc820799870
encoding/json.(*structEncoder).(encoding/json.encode)-fm(0xc820ace6e0, 0xc248c0, 0xc820a0ef00, 0x199, 0xc820a0ef00)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:601 +0x51 fp=0xc820799a50 sp=0xc820799a18
encoding/json.(*ptrEncoder).encode(0xc8203b4380, 0xc820ace6e0, 0xb4fac0, 0xc820996260, 0x96, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:709 +0xea fp=0xc820799aa0 sp=0xc820799a50
encoding/json.(*ptrEncoder).(encoding/json.encode)-fm(0xc820ace6e0, 0xb4fac0, 0xc820996260, 0x96, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:714 +0x51 fp=0xc820799ad8 sp=0xc820799aa0
encoding/json.(*structEncoder).encode(0xc820375c20, 0xc820ace6e0, 0xbaa6c0, 0xc820996240, 0x99, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:587 +0x2c4 fp=0xc820799c80 sp=0xc820799ad8
encoding/json.(*structEncoder).(encoding/json.encode)-fm(0xc820ace6e0, 0xbaa6c0, 0xc820996240, 0x99, 0xc820996200)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:601 +0x51 fp=0xc820799cb8 sp=0xc820799c80
encoding/json.(*encodeState).reflectValue(0xc820ace6e0, 0xbaa6c0, 0xc820996240, 0x99)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:301 +0x6b fp=0xc820799ce8 sp=0xc820799cb8
encoding/json.(*encodeState).marshal(0xc820ace6e0, 0xbaa6c0, 0xc820996240, 0x0, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:274 +0xa9 fp=0xc820799d30 sp=0xc820799ce8
encoding/json.Marshal(0xbaa6c0, 0xc820996240, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/encoding/json/encode.go:139 +0x84 fp=0xc820799d68 sp=0xc820799d30
github.com/flynn/flynn/pkg/sse.(*Stream).send(0xc820a18980, 0xbaa6c0, 0xc820996240, 0x0, 0x0)
	/home/ubuntu/go/src/github.com/flynn/flynn/pkg/sse/stream.go:117 +0xf2 fp=0xc820799e10 sp=0xc820799d68
github.com/flynn/flynn/pkg/sse.(*Stream).Serve.func2(0xc820a18980, 0xa3a580, 0xc8209e2840, 0x12, 0xa39e60, 0xc8209e2720, 0x12)
	/home/ubuntu/go/src/github.com/flynn/flynn/pkg/sse/stream.go:94 +0x265 fp=0xc820799f68 sp=0xc820799e10
runtime.goexit()
	/home/ubuntu/go/src/github.com/flynn/flynn/util/_toolchain/go/src/runtime/asm_amd64.s:1998 +0x1 fp=0xc820799f70 sp=0xc820799f68
created by github.com/flynn/flynn/pkg/sse.(*Stream).Serve
	/home/ubuntu/go/src/github.com/flynn/flynn/pkg/sse/stream.go:100 +0x3ca
```

**Comment by titanous** [comment]
This one is weird:

```text
t=2016-04-12T04:21:34+0000 lvl=info msg="changing state" app=containerinit fn=changeState state=running err= exitStatus=-1
t=2016-04-12T04:21:34+0000 lvl=info msg="monitoring service" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp
t=2016-04-12T04:21:34+0000 lvl=info msg="creating service" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp
t=2016-04-12T04:21:34+0000 lvl=info msg="sent state change" app=containerinit fn=StreamState state=running err= exitStatus=-1
t=2016-04-12T04:21:34+0000 lvl=info msg="adding healthcheck" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp type=tcp interval=0 threshold=0
t=2016-04-12T04:21:34+0000 lvl=warn msg="healthcheck error" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp component=monitor check=tcp://100.100.25.34:8080 err="dial tcp 100.100.25.34:8080: getsockopt: connection refused"
t=2016-04-12T04:21:34+0000 lvl=warn msg="healthcheck error" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp component=monitor check=tcp://100.100.25.34:8080 err="dial tcp 100.100.25.34:8080: getsockopt: connection refused"
t=2016-04-12T04:21:34+0000 lvl=warn msg="healthcheck error" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp component=monitor check=tcp://100.100.25.34:8080 err="dial tcp 100.100.25.34:8080: getsockopt: connection refused"
t=2016-04-12T04:21:34+0000 lvl=info msg="new monitor status" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp component=monitor status=up check=tcp://100.100.25.34:8080
t=2016-04-12T04:21:34+0000 lvl=info msg="created heartbeater" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp component=registrar addr=100.100.25.34:8080
runtime: checkdead: find g 126 in status 1
fatal error: checkdead: runnable g

runtime stack:
runtime.throw(0x7e227c)
	/usr/local/go/src/runtime/panic.go:491 +0xad
checkdead()
	/usr/local/go/src/runtime/proc.c:2828 +0x287
mput(0x7ed560)
	/usr/local/go/src/runtime/proc.c:3175 +0x47
stopm()
	/usr/local/go/src/runtime/proc.c:1176 +0xea
findrunnable(0xc208012000)
	/usr/local/go/src/runtime/proc.c:1487 +0x562
schedule()
	/usr/local/go/src/runtime/proc.c:1575 +0x151
runtime.park_m(0xc208000c60)
	/usr/local/go/src/runtime/proc.c:1654 +0x113
runtime.mcall(0x4336d4)
	/usr/local/go/src/runtime/asm_amd64.s:186 +0x5a

goroutine 1 [IO wait]:
net.(*pollDesc).Wait(0xc208010140, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208010140, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc2080100e0, 0x0, 0x7fbec7048b70, 0xc20806a780)
	/usr/local/go/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc208038018, 0x46b8fe, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:234 +0x4e
net/http.tcpKeepAliveListener.Accept(0xc208038018, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:1976 +0x4c
net/http.(*Server).Serve(0xc208044060, 0x7fbec7049e80, 0xc208038018, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:1728 +0x92
net/http.(*Server).ListenAndServe(0xc208044060, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:1718 +0x154
net/http.ListenAndServe(0xc20800a430, 0x5, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:1808 +0xba
main.main()
	/tmp/build/.heroku/go/src/github.com/flynn/flynn/test/apps/http/main.go:14 +0xc6

goroutine 126 [runnable]:
net.(*pollDesc).Wait(0xc208011c60, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208011c60, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc208011c00, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x7fbec7048b70, 0xc20806a7b0)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc208038210, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
net/http.(*liveSwitchReader).Read(0xc208046868, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:214 +0xab
io.(*LimitedReader).Read(0xc20801f380, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:408 +0xce
bufio.(*Reader).fill(0xc2080440c0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).ReadSlice(0xc2080440c0, 0xc20800a00a, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:295 +0x257
bufio.(*Reader).ReadLine(0xc2080440c0, 0x0, 0x0, 0x0, 0xc207ffc500, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:324 +0x62
net/textproto.(*Reader).readLineSlice(0xc20803ad50, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/textproto/reader.go:55 +0x9e
net/textproto.(*Reader).ReadLine(0xc20803ad50, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/textproto/reader.go:36 +0x4f
net/http.ReadRequest(0xc2080440c0, 0xc2080695f0, 0x0, 0x0)
	/usr/local/go/src/net/http/request.go:598 +0xcb
t=2016-04-12T04:25:36+0000 lvl=info msg="command exited" app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp status=2
net/http.(*conn).readRequest(0xc208046820, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/server.go:586 +0x26f
net/http.(*conn).serve(0xc208046820)
	/usr/local/go/src/net/http/server.go:1162 +0x69e
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:1751 +0x35e
t=2016-04-12T04:25:36+0000 lvl=info msg="received signal" app=containerinit fn=babySit type="child exited"
t=2016-04-12T04:25:36+0000 lvl=info msg="changing state" app=containerinit fn=changeState state=exited err= exitStatus=2
t=2016-04-12T04:25:36+0000 lvl=info msg=exiting app=containerinit fn=containerInitApp service=deploy-3b0aadbf-web port=8080 proto=tcp
t=2016-04-12T04:25:36+0000 lvl=info msg="sent state change" app=containerinit fn=StreamState state=exited err= exitStatus=2
```

**Comment by titanous** [comment]
@lmars I screwed up: I didn't upstream your fix for `TestMultipleHosts` and your recent changes make cc70a50d3bad1ecd07866dd4a4bcac53d5be31f6 fail to apply. If you fix it again I'll do my best to get this PR merged this week.

**Comment by lmars** [comment]
@titanous `TestMultipleHosts` fix: 6c00d9a

**Comment by titanous** [comment]
@lmars @josephglanville I think this is ready to go, can you give it one last look before I merge?

**Comment by josephglanville** [comment]
LGTM!  🎉

**Pull request merged**